src/main/java/com/github/danielwegener/logback/kafka/delivery/BaseDeliveryStrategy.java

```diff
@@ -0,0 +1,82 @@
+package com.github.danielwegener.logback.kafka.delivery;
+
+import ch.qos.logback.core.spi.ContextAwareBase;
+import org.apache.kafka.clients.producer.BufferExhaustedException;
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.errors.TimeoutException;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Future;
+
+/**
+ * Base implementation of {@link DeliveryStrategy}
+ * @since 0.0.1
+ */
+public abstract class BaseDeliveryStrategy<E, K, V> extends ContextAwareBase implements DeliveryStrategy<E, K, V> {
+    private HashMap<String, Object> producerConfig;
+    protected FailedDeliveryCallback<E> failedDeliveryCallback;
+
+    @Override
+    public void start(Map<String, Object> producerConfig, FailedDeliveryCallback<E> failedDeliveryCallback) {
+        this.producerConfig = new HashMap<>(producerConfig);
+        this.failedDeliveryCallback = failedDeliveryCallback;
+    }
+
+    /**
+     * Instantiates the producer with its config
+     * @return Created producer
+     * @throws KafkaException Producer initialization failed
+     */
+    protected Producer<K, V> createProducer() throws KafkaException{
+        return new KafkaProducer<>(new HashMap<>(producerConfig));
```

Comment:
why the copying? just as defense mechanism? or does the producer itself change the content of the config?

Reply:
yes defensive copy